src/2020-01-02-documenting-rcpp-packages.Rmd

```diff
@@ -0,0 +1,151 @@
+
+This vignette demonstrates four different ways to document Rcpp functions and classes in R packages. We will create an R package from scratch, add some very basic functionality and show how Roxygen2 blocks appear in the final documentation. Throughout the vignette, we will use several R packages to make our job easier: `devtools`, `usethis` and, of course, `rcpp`. All of these can be installed with the standard `install.packages()` interface:
+
+```{r eval=FALSE}
+install.packages( c("devtools", "usethis", "rcpp") )
+```
+
+## Create a package
+
+We begin by initializing a package skeleton. The following functions create a new directory and establish a barebones structure for an R package with Roxygen2 and Rcpp support:
+
+```{r eval=FALSE}
+devtools::create( "~/test/testpkg" )
+devtools::document()
+usethis::use_rcpp()
+```
+
+Using your favorite text editor, create a new file `~/test/testpkg/R/zzz.R` and add the following lines to it:
+
+```{r eval=FALSE}
+#' @useDynLib testpkg
+NULL
+
+Rcpp::loadModule("double_cpp", TRUE)
+```
+
+The `useDynLib` Roxygen block adds a NAMESPACE entry exposing the functionality of the dynamic library compiled from the .cpp source. Likewise `loadModule` exposes the functionality of an [Rcpp module](http://dirk.eddelbuettel.com/code/rcpp/Rcpp-modules.pdf) that we will use to encapsulate our C++ class. Replace `"double_cpp"` with your desired module name, but note that the name has to match what appears in the `.cpp` file below.
+
+## Add C++ functionality and Roxygen2 comment blocks
+
+All code blocks presented in this section are part of the same contiguous `.cpp` file. For the purposes of this tutorial, assume that the file lives in `~/test/testpkg/src/mult.cpp`. Let's review the four different ways to introduce Roxygen2 comment blocks inside `.cpp` files.
+
+### Approach 1: Standalone functions
+
+The first approach is the most straightforward and mimics traditional usage of Roxygen2 in pure R packages. The comments are placed immediately before a stand-alone function and begin with `//'`, which Rcpp will translate into `#'` in the corresponding `RcppExports.R` file:
+
+```{Rcpp eval=FALSE}
+#include <Rcpp.h>
+
+//' Multiplies two doubles
+//'
+//' @param v1 First value
+//' @param v2 Second value
+//' @return Product of v1 and v2
+//' @export
+// [[Rcpp::export]]
+double mult( double v1, double v2 ) {return v1*v2;}
+```
+
+You may notice two `export` keywords used in the comment block. The first, `@export`, tells Roxygen2 to add a `NAMESPACE` entry that will make the function visibile to the end users of your package. Remove this line if the function should be internal to the package.
```

Comment:
You may want to add that this is convenient, but optional.  I for one never use it in my packages.

Reply:
It's been a while since I looked at the details of this, but I thought that `@export` is required to make the function visible to the end user? Otherwise, `roxygenize()` doesn't add the necessary line in `NAMESPACE`?